docs/implementation.adoc

```diff
@@ -0,0 +1,507 @@
+
+* The `jenkins.scm.api.SCMSource` extension point is responsible for:
+
+** identifying heads / branches that are available
+
+** tracking the current revisions of heads
+
+** constructing `hudson.scm.SCM` instances for a given head and revision pair
+
+* The `jenkins.scm.api.SCMNavigator` extension point is responsible for:
+
+** enumerating potential `jenkins.scm.api.SCMSource` instances
+
+Some examples of how these concepts may be mapped into well known source control systems:
+
+* Git has a fairly direct mapping of concepts
+
+** `jenkins.scm.api.SCMHead` would represent the branches and tags in an individual Git repository
+
+** `jenkins.scm.api.SCMRevision` would represent the Git commit hash of each individual commit. Because Git commits are immutable, each `SCMRevision` is deterministic as if you ask to check out the same hash you will always get the same contents.
+
+** `jenkins.scm.api.SCMSource` would represent a single Git repository.
+
+** `jenkins.scm.api.SCMNavigator` would represent a collection of Git repositories - most likely all on the same server
+
+* Subversion can have a fairly direct mapping of concepts analogous to the Git mapping above.
+Alternatively some oganizations will use a single Subversion repository to hold multiple projects under which each project would have the familiar `trunk/branches/tags` structure. If we consider this style of Subversion use then:
+
+** `jenkins.scm.api.SCMHead` would be mapped against a specific project and be one of: `trunk`, `branches/_name_` or `tags/_name_`.
+
+** `jenkins.scm.api.SCMRevision` would be mapped to the global Subversion revision number against which the specific `SCMHead` was last modified. Each `SCMRevision` is deterministic as if you ask to check out the trunk/branch.tag at the same revision you will always get the same contents.
+
+** `jenkins.scm.api.SCMSource` would represent a single project withi the Subversion repository.
+
+** `jenkins.scm.api.SCMNavigator` would represent the collection of projects in the root of the Subversion repository.
+
+* CVS, one mapping of the CVS concepts would be:
+
+** `jenkins.scm.api.SCMHead` would represent a branch in a module.
+
+** `jenkins.scm.api.SCMRevision` would represent a timestamp. Sadly the timestamp is not deterministic as an immediate checkout may miss some files that are still being committed to the repository.
++
+NOTE: An alternative deterministic `jenkins.scm.api.SCMRevision` would be a list of all the files and their individual revisions, but this would be prohibitive in terms of storage cost and is non-trivial for a user to replicate from their own workspace and consequently the timestamp would be preferred.
+Ignoring the problems of time sychronization in a distributed system, we could assume that any timestamp more than _X_ minutes in the past should be deterministic in order to allow the consuming plugins to make optimizations that are only possible with deterministic revisions.
```

Comment:
A CVS timestamp is actually deterministic down to the granularity of one second. It is true that a developer might have been in the process of running a commit which took several minutes (this was in fact common when I worked on the NetBeans megarepo prior to Mercurial conversion), so the snapshot might be _incomplete_ (for example, uncompilable); but it is still very nearly _deterministic_ in that two different checkouts of the same `SCMRevision` should produce the same files—possibly differing only in some changes delivered within the span of one second, and even then only if one of the checkouts occurred _simultaneously_ with the commit, during that same second interval (checkouts happening later should agree on whether a given timestamp was before or after).